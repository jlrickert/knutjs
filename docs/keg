updated: 2024-06-17 03:50:44Z
kegv:    2023-01

title:   Knut documentation
url:     git@github.com:jlrickert/knut.git
creator: git@github.com:jlrickert/jlrickert.git
state:   living

summary: |
  Take control of your personal knowledge managment quickly by starting
  your own KEG with (Kn)owledge (Ut)ils.

indexes:
  - file: dex/nodes.tsv
    summary: all nodes by id
    plugin: nodes
  - file: dex/changes.md
    summary: latest changes
    plugin: changes
