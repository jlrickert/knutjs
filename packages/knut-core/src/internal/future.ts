import { Applicative1 } from 'fp-ts/lib/Applicative.js';
import { Apply1 } from 'fp-ts/lib/Apply.js';
import { Chain1 } from 'fp-ts/lib/Chain.js';
import { Functor1 } from 'fp-ts/lib/Functor.js';
import { Monad1 } from 'fp-ts/lib/Monad.js';
import { Pointed1 } from 'fp-ts/lib/Pointed.js';
import { identity, pipe } from 'fp-ts/lib/function.js';

export type Future<A> = Promise<A>;

export const URI = 'Future';
export type URI = typeof URI;
declare module 'fp-ts/HKT' {
	interface URItoKind<A> {
		readonly [URI]: Future<A>;
	}
}

export const future = {
	of<A>(a: A): Future<A> {
		return Promise.resolve(a);
	},
	ap<A>(fa: Future<A>): <B>(fab: Future<(a: A) => B>) => Future<B> {
		return (fab) => pipe(fab, this.apPar(fa));
	},
	apSeq<A>(fa: Future<A>): <B>(fab: Future<(a: A) => B>) => Future<B> {
		return (fab) => {
			return pipe(
				fab,
				this.chain((f) => pipe(fa, this.map(f))),
			);
		};
	},
	apPar<A>(fa: Future<A>): <B>(fab: Future<(a: A) => B>) => Future<B> {
		return (fab) => {
			return Promise.all([fab, fa]).then(([f, a]) => {
				return f(a);
			});
		};
	},
	tap<A>(f: (a: A) => void): (ma: Future<A>) => Future<A> {
		return (ma) => {
			return pipe(
				ma,
				this.map(f),
				this.chain(() => ma),
			);
		};
	},
	flatten<A>(a: Future<Future<A>>): Future<A> {
		return pipe(a, this.chain(identity));
	},
	map<A, B>(f: (a: A) => B): (fa: Future<A>) => Future<B> {
		return (fa) => {
			return fa.then(f);
		};
	},
	chain<A, B>(f: (a: A) => Future<B>): (fa: Future<A>) => Future<B> {
		return async (fa) => {
			const r = await f(await fa);
			return r;
		};
	},
	get Do(): Future<{}> {
		return this.of({});
	},
	bind<N extends string, A, B>(
		name: Exclude<N, keyof A>,
		f: (a: A) => Future<B>,
	): (ma: Future<A>) => Future<{
		readonly [K in keyof A | N]: K extends keyof A ? A[K] : B;
	}> {
		return async (ma) => {
			const value = f(await ma);
			return Object.assign({}, await ma, { [name]: value }) as any;
		};
	},
	bindTo<N extends string>(
		name: string,
	): <A>(ma: Future<A>) => Future<{ readonly [K in N]: A }> {
		return (ma) => {
			return pipe(
				this.Do,
				this.bind(name, () => ma),
			) as any;
		};
	},
	get Pointed(): Pointed1<URI> {
		return { URI: URI, of: this.of };
	},
	get Functor(): Functor1<URI> {
		return {
			URI: URI,
			map: this.map as any,
		};
	},
	get Monad(): Monad1<URI> {
		return {
			URI: URI,
			of: this.of,
			chain: (ma, fab) => this.chain(fab)(ma),
			map: (ma, fab) => this.map(fab)(ma),
			ap: (fab, ma) => this.apPar(ma)(fab),
		};
	},
	get Chain(): Chain1<URI> {
		return {
			URI: URI,
			chain: (ma, fab) => this.chain(fab)(ma),
			map: (ma, fab) => this.map(fab)(ma),
			ap: (fab, ma) => this.apPar(ma)(fab),
		};
	},
	get ApplyPar() {
		return {
			URI: URI,
			map: this.map,
			ap: this.apPar,
		};
	},
	get ApplySeq(): Apply1<URI> {
		return {
			URI: URI,
			map: (ma, fab) => this.map(fab)(ma),
			ap: (fab, ma) => this.apSeq(ma)(fab),
		};
	},
	get ApplicativePar(): Applicative1<URI> {
		return {
			URI: URI,
			map: (ma, fab) => this.map(fab)(ma),
			ap: (fab, ma) => this.apPar(ma)(fab),
			of: this.of,
		};
	},
	get ApplicativeSeq(): Applicative1<URI> {
		return {
			URI: URI,
			map: (ma, fab) => this.map(fab)(ma),
			ap: (fab, ma) => this.apSeq(ma)(fab),
			of: this.of,
		};
	},
};
