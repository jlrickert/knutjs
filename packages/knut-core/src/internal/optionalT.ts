import { Kind, URIS } from 'fp-ts/HKT';
import { Monad1 } from 'fp-ts/lib/Monad.js';
import { pipe } from 'fp-ts/lib/function.js';
import { Refinement } from 'fp-ts/lib/Refinement.js';
import { TOptional, optional } from './optional.js';
import { Predicate } from 'fp-ts/lib/Predicate.js';

export const some: <M extends URIS>(
	M: Monad1<M>,
) => <A>(a: NonNullable<A>) => Kind<M, TOptional<A>> = (M) => (a) =>
	pipe(a, optional.some, M.of);

export const none: <M extends URIS>(M: Monad1<M>) => Kind<M, TOptional<never>> = (M) =>
	M.of(optional.none as TOptional<never>);

export const alt: <M extends URIS>(
	M: Monad1<M>,
) => <B>(
	second: () => Kind<M, TOptional<B>>,
) => <A>(first: Kind<M, TOptional<A>>) => Kind<M, TOptional<A | B>> =
	(M) => (second) => (first) => {
		return M.chain(first, optional.match(second, some(M)) as any);
	};

export const zero: <M extends URIS>(M: Monad1<M>) => <A>() => Kind<M, TOptional<A>> =
	(M) => () => {
		return M.of(optional.none as any);
	};

export const fromNullable: <M extends URIS>(
	M: Monad1<M>,
) => <A>(a: A) => Kind<M, TOptional<NonNullable<A>>> = (M) => (a) => {
	return pipe(a, optional.fromNullable, M.of);
};

export const refine: <M extends URIS>(
	M: Monad1<M>,
) => <A, B extends A>(
	refinement: Refinement<A, B>,
) => (ma: Kind<M, TOptional<A>>) => Kind<M, TOptional<B>> =
	(M) => (refinement) => (ma) => {
		return pipe(
			ma,
			chain(M)(
				(a) => (refinement(a) ? some(M)(a as any) : none(M)) as any,
			),
		);
	};

export const filter: <M extends URIS>(
	M: Monad1<M>,
) => <A>(
	predicate: Predicate<A>,
) => (ma: Kind<M, TOptional<A>>) => Kind<M, TOptional<A>> =
	(M) => (predicate) => (ma) => {
		return pipe(
			ma,
			chain(M)(
				(a) => (predicate(a) ? some(M)(a as any) : none(M)) as any,
			),
		);
	};

export const match: <M extends URIS>(
	M: Monad1<M>,
) => <A, B, C>(
	onNone: () => C,
	onSome: (a: A) => B,
) => (ma: Kind<M, TOptional<A>>) => Kind<M, B | C> =
	(M) => (onNone, onSome) => (ma) => {
		return M.map(ma, optional.match(onNone, onSome));
	};

export const map: <M extends URIS>(
	M: Monad1<M>,
) => <A, B>(
	f: (a: A) => B,
) => (ma: Kind<M, TOptional<A>>) => Kind<M, TOptional<B>> =
	(M) => (f) => (ma) => {
		return M.map(ma, optional.map(f));
	};

export const chain: <M extends URIS>(
	M: Monad1<M>,
) => <A, B>(
	f: (a: A) => Kind<M, TOptional<B>>,
) => (ma: Kind<M, TOptional<A>>) => Kind<M, TOptional<B>> =
	(M) => (f) => (ma) => {
		return M.chain(
			ma,
			optional.match(() => zero(M)(), f),
		);
	};

export const ap: <M extends URIS>(
	M: Monad1<M>,
) => <A>(
	ma: Kind<M, TOptional<A>>,
) => <B>(fab: Kind<M, TOptional<(a: A) => B>>) => Kind<M, TOptional<B>> =
	(M) => (ma) => (mab) => {
		const x = M.ap(
			M.map(mab, (gab) => (ga: any) => optional.ap(ga)(gab)),
			ma,
		);
		return x;
	};

export const getOrElse: <M extends URIS>(
	M: Monad1<M>,
) => <A>(
	onNone: () => Kind<M, A>,
) => (ma: Kind<M, TOptional<A>>) => Kind<M, A> = (M) => (onNone) => (ma) => {
	return M.chain(ma, optional.match(onNone, M.of));
};

export const Do: <M extends URIS>(M: Monad1<M>) => Kind<M, TOptional<{}>> = (M) =>
	some(M)({});

export const bind: <M extends URIS>(
	M: Monad1<M>,
) => <N extends string, A, B>(
	name: Exclude<N, keyof A>,
	f: (a: A) => Kind<M, TOptional<B>>,
) => (
	ma: Kind<M, TOptional<A>>,
) => Kind<
	M,
	TOptional<{ readonly [K in keyof A | N]: K extends keyof A ? A[K] : B }>
> = (M) => (name, f) => (ma) => {
	return M.chain(
		ma,
		optional.match(
			() => zero(M)(),
			(a) => {
				const mb = f(a);
				return M.chain(
					mb,
					optional.map(
						(b) => Object.assign({}, a, { [name]: b }) as any,
					),
				);
			},
		),
	);
};

export const bindTo: <M extends URIS>(
	M: Monad1<M>,
) => <N extends string>(
	name: N,
) => <A>(
	ma: Kind<M, TOptional<A>>,
) => Kind<M, TOptional<{ readonly [K in N]: A }>> = (M) => (name) => (ma) => {
	return pipe(
		Do(M),
		bind(M)(name, () => ma),
	) as any;
};

export const optionalT = <M extends URIS>(M: Monad1<M>) => ({
	some: some(M),
	none: none(M),
	alt: alt(M),
	zero: zero(M),
	fromNullable: fromNullable(M),
	refine: refine(M),
	filter: filter(M),
	match: match(M),
	ap: ap(M),
	map: map(M),
	chain: chain(M),
	getOrElse: getOrElse(M),
	Do: Do(M),
	bind: bind(M),
	bindTo: bindTo(M),
});
