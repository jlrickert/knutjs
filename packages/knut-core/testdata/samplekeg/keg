updated: 2022-11-26 19:33:24Z
kegv:    2023-01

title:   A Sample Keg
url:     git@github.com:YOU/keg.git
creator: git@github.com:YOU/YOU.git
state:   living

summary:
  ðŸ‘‹ Hey there! The KEG community welcomes you. This is an initial
  sample `keg` file. It uses a simplified YAML format. (If you know JSON,
  you know this.) Go ahead and change this summary and anything else you
  like, but keep in mind the following:
  
  * The updated line MUST be the first line
  * Everything *but* the updated line is optional
  * You can leave the indexes section alone (they are on by default)
  * Change **creator** to the URL for your main keg
  * Change the **url** to the main place to find this keg
  * Change **state** only if you are planning something more formal
  * Keep the **zero node** created to link to for planned content
  
  If you are a `vim` user you might want to add something like the
  following to your `.vimrc`:
  
    au bufnewfile,bufRead keg set ft=yaml

indexes:
  - file: dex/changes.md
    summary: latest changes
  - file: dex/nodes.tsv
    summary: all nodes by id