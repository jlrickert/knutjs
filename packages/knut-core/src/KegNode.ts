import { pipe } from 'fp-ts/lib/function.js';
import { MetaData, MetaFile } from './metaFile.js';
import { stringify } from './utils.js';
import { NodeContent } from './nodeContent.js';
import { TStorage } from './storage/Storage.js';
import { Future } from './internal/future.js';
import { Optional } from './internal/index.js';

export type NodeData = {
	content: NodeContent;
	updated: Date;
	metaData: MetaData;
};

export type NodeOptions = {
	content: string;
	updated?: Date;
	meta?: MetaData;
};

export type NodeId = number;

/**
 * Node represents a handle to a node. For immutable see NodeData
 */
export class KegNode {
	/**
	 * Look in the path for a number.  First number found is returned
	 */
	static parseNodeId(path: string): Optional.TOptional<number> {
		const parts = path.split('/');
		for (const part of parts) {
			const id = Number(part);
			if (!Number.isNaN(id)) {
				return id;
			}
		}
		return null;
	}

	static isNode(obj: any): obj is KegNode {
		return obj instanceof KegNode;
	}

	static async fromStorage(
		id: NodeId,
		storage: TStorage,
	): Future<Optional.TOptional<KegNode>> {
		const nodePath = NodeContent.filePath(id);
		const metaPath = MetaFile.filePath(id);
		const rawContent = await storage.read(nodePath);
		if (Optional.isNone(rawContent)) {
			return Optional.none;
		}
		const yamlData = await storage.read(metaPath);
		const metaData = yamlData
			? MetaFile.fromYAML(yamlData)
			: new MetaFile();
		const stats = await storage.stats(nodePath);
		if (Optional.isNone(rawContent) && Optional.isNone(stats)) {
			return Optional.none;
		}
		const content = await NodeContent.fromMarkdown(rawContent);
		const node = new KegNode(id, {
			content,
			updated: stats?.mtime ? new Date(stats.mtime) : new Date(),
			metaData: metaData.data,
		});
		return node;
	}

	static readmePath(nodeId: number): string {
		return `${nodeId}/README.md`;
	}

	static metaPath(nodeId: NodeId): string {
		return `${nodeId}/meta.yaml`;
	}

	static async fromContent(options: NodeOptions): Future<NodeData> {
		const content = await NodeContent.fromMarkdown(options.content);
		return {
			content,
			metaData: options.meta ?? new MetaFile().data,
			updated: pipe(
				options.updated,
				Optional.getOrElse(() => new Date()),
			),
		};
	}

	static async zeroNode(): Future<NodeData> {
		const content = await NodeContent.fromMarkdown(
			`
# Sorry, planned but not yet available

This is a filler until I can provide someone better for the link that brought you here. If you are really anxious, consider opening an issue describing why you would like this missing content created before the rest.
				`.trim(),
		);
		const now = new Date();
		const meta = new MetaFile();
		return { content, metaData: meta.data, updated: now };
	}

	static fromData(id: NodeId, data: NodeData): KegNode {
		return new KegNode(id, data);
	}

	public constructor(
		public id: number,
		private data: NodeData,
	) {}

	get title(): string {
		return this.data.content.title ?? '';
	}

	set title(value: string) {
		this.data.content.title = value;
	}

	get content(): string {
		return stringify(this.data.content);
	}

	get meta(): MetaFile {
		return new MetaFile(this.data.metaData);
	}

	get updated(): Date {
		return this.data.updated;
	}

	set updated(date: Date) {
		this.data.updated = date;
	}

	async updateContent(content: string, now?: Date): Future<void> {
		this.data.content = await NodeContent.fromMarkdown(content);
		this.updated = now ?? new Date();
	}

	async toStorage(storage: TStorage): Future<boolean> {
		const ok = [
			await storage.write(KegNode.readmePath(this.id), this.content),
			await storage.utime(KegNode.readmePath(this.id), {
				mtime: this.updated,
			}),
			await storage.write(
				KegNode.metaPath(this.id),
				stringify(this.meta),
			),
		].reduce((a, b) => a || b, false);

		return ok;
	}
}
