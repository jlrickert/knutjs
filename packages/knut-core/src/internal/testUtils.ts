import * as Path from 'path';
import invariant from 'tiny-invariant';
import { mkdtemp, rm } from 'fs/promises';
import { tmpdir } from 'os';
import { afterEach, describe } from 'vitest';
import { optional } from './optional.js';
import { Future } from './future.js';
import { Loader, Backend } from '../backend.js';
import { overwrite } from '../storage/storageUtils.js';
import { GenericStorage } from '../storage/storage.js';
import { NodeStorage } from '../storage/nodeStorage.js';
import { WebStorage } from '../storage/webStorage.js';

export type Kegpath = 'samplekeg1' | 'samplekeg2' | 'samplekeg3';

const fixturePath = Path.resolve(__dirname, '..', '..', 'testdata');

/**
 * Path to sample keg fixture
 **/
const fixtureKegpath = (Kegpath: Kegpath) => Path.resolve(fixturePath, Kegpath);

/**
 * Config path to the config test fixtures
 **/
const FIXTURE_CONFIG_PATH = Path.resolve(fixturePath, 'config/knut');

/**
 * Variable path to fixtures
 **/
const FIXTURE_VAR_PATH = Path.resolve(fixturePath, 'local/share/knut');

/**
 * Path to cache fixture
 **/
const FIXTURE_CACHE_PATH = Path.resolve(fixturePath, 'cache/knut');

const fixtureStorage = new NodeStorage(fixturePath);

const tempNodeStorage: () => Future<GenericStorage> = async () => {
	const rootPath = await mkdtemp(Path.join(tmpdir(), 'knut-test-'));
	const storage = new NodeStorage(rootPath);

	afterEach(async () => {
		try {
			await rm(storage.root, { recursive: true });
		} catch (e) {}
	});

	return storage;
};

const testBrowserBackend = async (): Future<Backend> => {
	const fixture = fixtureStorage;

	const storage = WebStorage.create('knut');

	// Child path needs to match up with fixture cache
	const cache = storage.child('cache/knut');

	// Variable path needs to match up with fixture variable path
	const variable = storage.child('local/share/knut');

	// Config path needs to match up with fixture path so that config.yaml
	// resolves to the correct location
	const config = storage.child('config/knut');

	const kegStorage = WebStorage.create('knut-kegs');
	const loader: Loader = async (uri: string) => {
		const storage = kegStorage.child(uri);
		return storage;
	};

	const backend = {
		config,
		variable,
		cache,
		loader,
	} satisfies Backend;

	for (const kegalias of [
		'samplekeg1',
		'samplekeg2',
		'samplekeg3',
	] satisfies Kegpath[]) {
		const storage = await backend.loader(kegalias);
		const path = fixtureKegpath(kegalias);
		invariant(
			optional.isSome(storage),
			`Expect fixture to load keg storage`,
		);
		await overwrite(fixture.child(path), storage);
	}

	await overwrite(fixture.child(FIXTURE_CONFIG_PATH), backend.config);
	await overwrite(fixture.child(FIXTURE_VAR_PATH), backend.variable);
	await overwrite(fixture.child(FIXTURE_CACHE_PATH), backend.cache);
	return backend;
};

const testEmptyNodeBackend = async (): Future<Backend> => {
	const root = await tempNodeStorage();

	const cache = root.child('cache/knut');
	const config = root.child('config/knut');
	const variable = root.child('local/share/knut');

	const loader: Loader = async (uri) => {
		return root.child(uri);
	};

	return {
		config,
		variable,
		cache,
		loader,
	};
};

/**
 * Returns a platform with full fixtures
 **/
const testNodeBackend = async (): Future<Backend> => {
	const root = await tempNodeStorage();

	const cache = root.child('cache/knut');
	const config = root.child('config/knut');
	const variable = root.child('local/share/knut');
	await overwrite(fixtureStorage, root);

	const loader: Loader = async (uri) => {
		return root.child(uri);
	};

	return {
		config,
		variable,
		cache,
		loader,
	};
};

const backends = [
	{
		name: 'Node',
		getBackend: testNodeBackend,
	},
	{
		name: 'Browser',
		getBackend: testBrowserBackend,
	},
] as const;

export const testUtils = {
	tempNodeStorage,
	testNodeBackend,
	testEmptyNodeBackend,
	testBrowserBackend,
	fixtureStorage,
	backends,
};

/**
 * TestUtils is a namespace for utilities to make testing much easier
 */
export class TestUtils {
	static readonly backends = [
		{
			name: 'Node',
			loadBackend: testNodeBackend,
		},
		{
			name: 'Browser',
			loadBackend: testBrowserBackend,
		},
	];

	/**
	 * fixtureStorage is the raw data for fixtures.
	 * DO NOT WRITE ANYTHING HERE!!!
	 **/
	static readonly fixtureStorage = fixtureStorage;

	/**
	 * makeNodeBackend creates a temporary directory loaded with fixtures.
	 * This is intened for testing when a node environment is available.
	 */
	static readonly makeNodeBackend = testNodeBackend;

	/**
	 * makeEmptyNodeBackend creates a temporary directory loaded with fixtures.
	 * This is intened for testing when a node environment is available.
	 */
	static readonly makeEmptyNodeBackend = testEmptyNodeBackend;

	/**
	 * makeBrowserBackend creates backend for testing storage on a browser.
	 */
	static readonly makeBrowserBackend = testBrowserBackend;

	/**
	 * makeTempstorage creates a temporary storage in memory.
	 *
	 * On mac I think this creates a new direcory on a ram disk
	 */
	static readonly makeTempStorage = tempNodeStorage;

	static readonly describeEachBackend = async (
		desc: string,
		factory: (f: (typeof TestUtils)['backends'][number]) => Future<void>,
	) => {
		for await (const { name, loadBackend } of TestUtils.backends) {
			describe(`${desc} (${name})`, async () => {
				factory({ name, loadBackend });
			});
		}
	};
}
