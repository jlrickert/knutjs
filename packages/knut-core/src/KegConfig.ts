import { Array, Option } from 'effect';
import { pipe } from 'fp-ts/lib/function.js';
import * as YAML from 'yaml';
import { Json, deepCopy, stringify } from './utils.js';
import { NodeId } from './KegNode.js';
import { TStorage } from './storage/Storage.js';
import { Future, Optional } from './internal/index.js';

export type KegVersion = '2023-01';
const defaultSchema =
	'https://raw.githubusercontent.com/jlrickert/knutjs/main/packages/knut-core/kegSchema.json';

/**
 * Plain old data representing an Index. This data is found in the **keg**
 * file.
 **/
export type IndexEntryData = {
	/**
	 * file relative to the keg file. By convention this will point to a file
	 * in the dex directory.
	 */
	file: string;

	/**
	 * Summary for the index
	 **/
	summary?: string;

	/**
	 * Name of the plugin to use.
	 **/
	plugin?: string;

	args?: Json;
};

export type KegPluginData = {
	/**
	 * Name of the plugin
	 */
	name: string;
	/**
	 * Whether the plugin is enabled or not
	 */
	enable: boolean;
};

/**
 * Plain old data representing the data in the *keg* file. This should match
 * kegSchema.json file
 */
export type KegConfigData = {
	$schema: typeof defaultSchema;
	/**
	 * last time the index has been updated.
	 **/
	updated?: Date;
	kegv?: KegVersion;
	title?: string;
	url?: string;
	linkfmt?: string;
	creator?: string;
	state?: string;
	summary?: string;
	format?: 'json' | 'yaml';
	plugins?: KegPluginData[];
	indexes?: IndexEntryData[];
};

const defaultKegConfigData = (updated?: Date): KegConfigData => ({
	$schema: defaultSchema,
	updated: updated ?? new Date(),
	kegv: '2023-01',
	state: 'living',
	title: 'An keg',
	url: 'git@github.com:YOU/keg.git',
	creator: 'git@github.com:YOU/YOU.git',
	plugins: [
		{ name: 'date', enable: true },
		{ name: 'tags', enable: true },
	],
	indexes: [
		{
			file: 'dex/nodes.tsv',
			summary: 'all nodes by id',
		},
		{
			plugin: 'date',
			file: 'dex/changes.md',
			summary: 'latest changes',
		},
		{
			plugin: 'tags',
			file: 'dex/tags.md',
			summary: 'all tags',
		},
	],
});

/**
 * Keg config service for managing configuration for a keg
 */
export class KegConfig {
	/**
	 * Load a keg file for the given path
	 */
	static fromStorage = async (
		storage: TStorage,
	): Future.Future<Optional.Optional<KegConfig>> => {
		return pipe(
			await storage.read('keg'),
			Optional.map((a) => KegConfig.fromYAML(a)),
		);
	};

	static fromYAML = (yaml: string): KegConfig => {
		const data = YAML.parse(yaml) as KegConfigData;
		return new KegConfig(data);
	};

	static fromJSON = (json: string): KegConfig => {
		const data = JSON.parse(json);
		return new KegConfig(data);
	};

	static default = (): KegConfig => {
		return new KegConfig({});
	};

	public data: KegConfigData;

	public constructor(
		/**
		 * Raw data for KegConfigData
		 */
		data: Partial<KegConfigData>,
	) {
		this.data = defaultKegConfigData();
		this.mergeData(data);
	}

	clone() {
		return new KegConfig(deepCopy(this.data));
	}

	mergeData(data: Partial<KegConfigData>) {
		const plugins = pipe(
			Array.dedupeWith(
				[...(this.data.plugins ?? []), ...(data.plugins ?? [])],
				(a, b) => a.name === b.name,
			),
		);
		const indexes = pipe(
			Array.dedupeWith(
				[...(this.data.indexes ?? []), ...(data.indexes ?? [])],
				(a, b) => a.file === b.file,
			),
		);
		this.data = { ...this.data, ...data, plugins, indexes };
	}

	*indexIter() {
		const entries: IndexEntryData[] = [];
		for (const entry of this.data.indexes ?? []) {
			entries.push(entry);
			yield entry satisfies IndexEntryData;
		}
		return entries;
	}

	get format(): KegConfigData['format'] {
		if (Optional.isNone(this.data.format)) {
			return 'yaml';
		}
		if (['yaml', 'json'].includes(this.data.format)) {
			return this.data.format;
		}
		return 'yaml';
	}

	set format(format: 'yaml' | 'json') {
		this.data.format = format;
	}

	/**
	 * Looks up plugin configuration details for a plugin
	 */
	lookupPluginData(plugin: string) {
		const isEnabled = pipe(
			this.data.plugins ?? [],
			Array.findFirst((a) => a.name === plugin),
			Option.match({ onNone: () => false, onSome: (a) => a.enable }),
		);
		const indexList = pipe(
			this.data.indexes ?? [],
			Array.filter((a) => a.plugin === plugin),
			Array.dedupeWith(
				(a, b) => a.plugin === b.plugin && a.file === b.file,
			),
		);

		return {
			isEnabled,
			indexList,
		};
	}

	async toStorage(
		storage: TStorage,
		options?: { updateTimeStamp?: boolean },
	): Future.Future<boolean> {
		if (options?.updateTimeStamp) {
			this.mergeData({ updated: new Date() });
		}
		return await storage.write('keg', stringify(this));
	}

	getAuthor(): Optional.Optional<string> {
		return this.data.creator ?? null;
	}

	getLink(nodeId: NodeId): Optional.Optional<string> {
		const linkfmt = this.data.linkfmt;
		return linkfmt ? linkfmt.replace('{{id}}', nodeId.stringify()) : null;
	}

	toYAML(): string {
		const schemaLine = `# yaml-language-server: $schema=${defaultSchema}`;
		const content = YAML.stringify(this.data);
		return [schemaLine, content].join('\n');
	}

	toJSON(pretty?: boolean): string {
		if (pretty) {
			return JSON.stringify(this.data, undefined, 4);
		}
		return JSON.stringify(this.data);
	}

	stringify(): string {
		return this.toYAML();
	}
}
