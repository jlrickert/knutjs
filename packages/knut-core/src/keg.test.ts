import { pipe } from 'fp-ts/lib/function.js';
import { afterEach, describe, expect, it, vi } from 'vitest';
import invariant from 'tiny-invariant';
import { TestUtils } from './internal/testUtils.js';
import { future } from './internal/future.js';
import { optionalT } from './internal/optionalT.js';
import { Keg } from './keg.js';
import { deepCopy, stringify } from './utils.js';
import { KegNode, NodeId } from './KegNode.js';
import { DexEntry } from './Dex.js';
import { KegConfig } from './KegConfig.js';
import { Optional } from './internal/index.js';

const T = optionalT(future.Monad);

TestUtils.describeEachBackend('Keg', async ({ name, loadBackend }) => {
	afterEach(() => {
		global.localStorage.clear();
		vi.resetAllMocks();
	});

	it('should be able to create a new keg', async () => {
		const backend = await loadBackend();
		const storage = await backend.loader('testkeg');
		invariant(
			Optional.isSome(storage),
			'Expect test backend to load as expected',
		);
		const keg = await Keg.createNew(storage);
		invariant(Optional.isSome(keg));

		const kegFileContent = await storage.read('keg');
		expect(kegFileContent).toEqual(stringify(keg?.config));

		const node = await keg.getNode(new NodeId(0));
		const zeroNode = await KegNode.zeroNode();
		expect(node?.content).toEqual(zeroNode.content);
	});

	it('should be able to create a new node', async () => {
		const backend = await loadBackend();
		const storage = await backend.loader('testkeg');
		invariant(
			Optional.isSome(storage),
			'Expect test backend to load as expected',
		);
		const keg = await Keg.createNew(storage);
		invariant(Optional.isSome(keg));

		const nodeId = await keg.createNode();
		const node = await pipe(
			future.of(nodeId),
			T.chain((id) => keg.getNode(id)),
		);

		expect(node).toBeInstanceOf(KegNode);
	});

	it('should be able to edit a node', async () => {
		const now = new Date('2023-03-23');
		vi.useFakeTimers();
		vi.setSystemTime(now);

		const backend = await loadBackend();
		const storage = await backend.loader('testkeg2');
		invariant(
			Optional.isSome(storage),
			'Expect test backend to load as expected',
		);
		const keg = await Keg.createNew(storage);
		invariant(Optional.isSome(keg));
		const id = 2;
		await pipe(
			KegNode.fromContent({ content: '# Sample node', updated: now }),
			future.chain((node) => keg.writeNode(new NodeId(id), node)),
		);

		expect(
			await pipe(
				keg.getNode(new NodeId(id)),
				T.map((node) => ({
					content: node.content,
					updated: node.updated,
				})),
			),
		).toStrictEqual({ content: '# Sample node\n', updated: now });
	});

	describe(`${name} - keg indexing`, async () => {
		afterEach(() => {
			global.localStorage.clear();
			vi.resetAllMocks();
		});

		it('should be able to rebuild the dex index', async () => {
			const backend = await loadBackend();
			const storage = await backend.loader('testkeg');
			invariant(
				Optional.isSome(storage),
				'Expect test backend to load as expected',
			);

			const now = new Date('2023-03-23');
			vi.useFakeTimers();
			vi.setSystemTime(now);

			const keg = await Keg.createNew(storage);
			invariant(Optional.isSome(keg));

			const zeroNode = await KegNode.zeroNode();
			expect(keg.dex.entries).toHaveLength(1);
			expect(keg.dex.entries).toStrictEqual([
				expect.objectContaining<Partial<DexEntry>>({
					title: zeroNode.title,
					updated: now,
				}),
			]);
			const nodeId = await keg.createNode();
			invariant(Optional.isSome(nodeId));

			const later = new Date(new Date(now).setMinutes(30));
			vi.setSystemTime(later);

			const content = '# New node\nExample summary';
			await keg.writeNodeContent(nodeId, {
				content,
			});

			expect(keg.dex.entries).toHaveLength(2);
			expect(keg.dex.entries).toStrictEqual([
				expect.objectContaining({
					title: zeroNode.title,
					updated: now,
				}),
				expect.objectContaining({
					title: 'New node',
					updated: later,
				}),
			]);

			expect(await storage.read('dex/nodes.tsv')).matchSnapshot();
		});

		it('should create a daily.md with valid date plugin config', async () => {
			const backend = await loadBackend();
			const storage = await backend.loader('testkeg');
			invariant(
				Optional.isSome(storage),
				'Expect test backend to load as expected',
			);

			const now = new Date('2023-03-23');
			vi.useFakeTimers();
			vi.setSystemTime(now);

			const keg = await Keg.createNew(storage);
			invariant(Optional.isSome(keg));

			const config = new KegConfig(deepCopy(keg.config.data));
			config.mergeData({
				indexes: [
					{
						plugin: 'date',
						file: 'dex/daily.md',
						args: { tags: 'daily' },
						summary: 'Daily index',
					},
				],
			});
			await keg.reloadConfig(config);
			invariant(Optional.isSome(keg));

			const [a, b, c, d, e] = await Promise.all([
				keg.createNode(),
				keg.createNode(),
				keg.createNode(),
				keg.createNode(),
				keg.createNode(),
			]);
			invariant(Optional.isSome(a));
			invariant(Optional.isSome(b));
			invariant(Optional.isSome(c));
			invariant(Optional.isSome(d));
			invariant(Optional.isSome(e));

			await keg.writeNodeContent(a, {
				content: '# Node A',
			});
			await keg.writeNodeContent(b, {
				content: '# Node B',
				meta: { tags: ['daily'], date: stringify(now) },
			});
			await keg.writeNodeContent(c, {
				content: '# Node C',
				meta: { date: stringify(now) },
			});
			await keg.writeNodeContent(d, {
				content: '# Node D',
				meta: { tags: ['daily'] },
			});
			await keg.writeNodeContent(e, {
				content: '# Node E',
				meta: { tags: ['other'] },
			});

			expect(await keg.storage.read('dex/daily.md')).matchSnapshot();
		});
	});
});
