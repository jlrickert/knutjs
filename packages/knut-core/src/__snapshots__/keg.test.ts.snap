// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Browser - keg indexing > should be able to rebuild the index 1`] = `
"0	2023-03-23T00:00:00.000Z	Sorry, planned but not yet available
1	2023-03-23T00:30:00.000Z	New node"
`;

exports[`Browser - keg indexing > should create a daily.md with valid date plugin config 1`] = `null`;

exports[`Keg (Browser) > Browser - keg indexing > should be able to rebuild the dex index 1`] = `"0	2023-03-23T00:00:00.000Z	Sorry, planned but not yet available"`;

exports[`Keg (Browser) > Browser - keg indexing > should create a daily.md with valid date plugin config 1`] = `null`;

exports[`Keg (Node) > Node - keg indexing > should be able to rebuild the dex index 1`] = `"0	2023-03-23T00:00:00.000Z	Sorry, planned but not yet available"`;

exports[`Keg (Node) > Node - keg indexing > should create a daily.md with valid date plugin config 1`] = `null`;

exports[`Node - keg indexing > should be able to rebuild the index 1`] = `
"0	2023-03-23T00:00:00.000Z	Sorry, planned but not yet available
1	2023-03-23T00:30:00.000Z	New node"
`;

exports[`Node - keg indexing > should create a daily.md with valid date plugin config 1`] = `null`;
